docs/architecture/psa-thread-safety.md

```diff
@@ -293,8 +293,72 @@ There is currently no indication of when a slot is in the WRITING state. This on
+
+1. Depend on the deletion marker. If the slot has been reused and is marked for deletion again, the threads keep waiting until the second deletion completes.
+2. Introduce a uuid (eg a global counter plus a slot ID), which is recorded by the thread waiting for deletion and checks whether it matches. If it doesn't, the function can return as the slot was already reallocated. If it does match, it can check whether it is still marked for deletion, if it is, the thread goes back to sleep, if it isn't, the function can return.
+
+#### Platform abstraction
+
+Introducing condition variables to the platform abstraction layer would be best done in a major version. If we can't wait until that, we will need to introduce a new compile time flag. Considering that this only will be needed on the PSA Crypto side and the upcoming split, it makes sense to make this flag responsible for the entire PSA Crypto threading support. Therefore if we want to keep the option open for implementing this in a backward compatible manner, we need to introduce and use this new flag already when implementing #mutex-only. (If we keep the abstraction layer for mutexes the same, this shouldn't mean increase in code size and would mean only minimal effort on the porting side.)
+
+### Operation contexts
+
+Concurrent access to the same operation context can compromise the crypto service for example if the operation context has a pointer (depending on the compiler and the platform, the pointer assignment may or may not be atomic). This violates the functional correctness requirement. (Concurrent calls to operations is undefined behaviour, but still should not compromise the CIA of the crypto service.)
+
+Operations will have a status field protected by a global mutex similarly to key slots. On entry, API calls check the state and return an error if it is already ACTIVE. Otherwise they set it to ACTIVE and restore it to INACTIVE before returning.
+
+### Drivers
+
+Each driver that hasn’t got the "thread_safe” property set  has a dedicated mutex.
+
+Implementing "thread_safe” drivers depends on the condition variable protection in the key store, as we must guarantee that the core never starts the destruction of a key while there are operations in progress on it.
+
+Start with implementing threading for drivers without the "thread_safe” property (all drivers behave like the property wasn't set). Add "thread_safe" drivers at some point after the #condition-variables approach is implemented in the core.
+
+### Global Data
+
+PSA Crypto makes use of a `global_data` variable that will be accessible from multiple threads and needs to be protected. Any function accessing this variable (or its members) must take the corresponding lock first. Since `global_data` holds the RNG state, these will involve relatively expensive operations and therefore ideally `global_data` should be protected by its own, dedicated lock (different from the one protecting the key store).
```

Comment:
It is enough, if we really want to, as everything else is a bitfield and is written unconditionally and only once. But we should document this so that we don't break it when we modify it in the future.

Reply:
Ok, I was wrong. Everything else is a bitfield and is written unconditionally and only once, that is true. This is indeed enough to guarantee that there won't be race conditions when we consider the fields individually. However, if we want to guarantee that the whole structure is in a consistent state, we will need to protect it with a mutex. (For example, if `mbedtls_psa_crypto_free` and `mbedtls_psa_crypto_init` are called in two different threads, it can leave `global_data` in an inconsistent state.)

Reply:
Update: We can't make `mbedtls_psa_crypto_free()` thread safe at this time. Also, it is out of scope as it is not a PSA Crypto API, it is an Mbed TLS specific extension. We still need some locking mechanism to allow parallel init (as init can't busy-wait and can't exit without the initialisation being complete either).